### PR TITLE
New RPCs for Push and Download

### PR DESCRIPTION
**Pull request by unmultimedio**
## ⚠️ This is just a proof of concept to have something that compiles to use it in the backend.

To enrich existing conversation on new RPCs for tamper proofing abilities.

Fixes: TCN-905

### CODE REVIEW

File: proto/buf/alpha/registry/v1alpha1/download.proto [hunk 1]
```diff
@@ -21,20 +21,38 @@ import "buf/alpha/module/v1alpha1/module.proto";
 // DownloadService is the download service.
 service DownloadService {
   // Download downloads.
+  // Soon to be deprecated.
```
Comment (pkwarren) [inline proto/buf/alpha/registry/v1alpha1/download.proto:24]: We should state it is soon to be deprecated in favor of DownloadManifest.

File: proto/buf/alpha/registry/v1alpha1/download.proto [hunk 2]
```diff
@@ -21,20 +21,38 @@ import "buf/alpha/module/v1alpha1/module.proto";
...
   rpc Download(DownloadRequest) returns (DownloadResponse);
+  // DownloadManifest downloads a module in the manifest+blobs encoding format.
+  rpc DownloadManifest(DownloadManifestRequest) returns (DownloadManifestResponse);
```
Comment (pkwarren) [inline proto/buf/alpha/registry/v1alpha1/download.proto:27]: Do we want to name these `DownloadManifestAndBlobs` to be more accurate?

**Review by pkwarren — COMMENTED** [review]
Looks good. For the build to go green we'll need to regenerate the code.

File: proto/buf/alpha/registry/v1alpha1/download.proto (hunk 2 above)
Comment (unmultimedio) [reply proto/buf/alpha/registry/v1alpha1/download.proto:27]: Yes could be, I'm more for shorter, so another option is `*CAS` suffix, but that could be cryptic?

File: proto/buf/alpha/registry/v1alpha1/download.proto (hunk 1 above)
Comment (unmultimedio) [reply proto/buf/alpha/registry/v1alpha1/download.proto:24]: Sure, we could expand on _why_ we're not deprecating it _yet_.

File: proto/buf/alpha/registry/v1alpha1/download.proto (hunk 2 above)
Comment (pkwarren) [reply proto/buf/alpha/registry/v1alpha1/download.proto:27]: I don't mind *CAS.

**Comment by bufdev** [comment]
Can you expand on why we're making new RPCs for this?

**Comment by bufdev** [comment]
If there's a need to change the behavior of `Download`, the cleanest way to do this would be to have a new `buf.alpha.registry.v1alpha2` package with a new `DownloadService`, as opposed to a new RPC. This may be prohibitive in terms of time to market, as a lot of stuff is organized around `buf.alpha.registry.v1alpha1`, but just of note, this _is_ adding tech debt.

**Comment by unmultimedio** [comment]
> Can you expand on why we're making new RPCs for this?

This is to support tamper proofing abilities in an upcoming CLI version that will push/download blobs/files with a consistent encoding instead of modules in a proto message. I will share the document internally!

File: proto/buf/alpha/registry/v1alpha1/download.proto [hunk 3]
```diff
@@ -21,20 +21,38 @@ import "buf/alpha/module/v1alpha1/module.proto";
 // DownloadService is the download service.
 service DownloadService {
   // Download downloads.
+  // Soon to be deprecated.
   rpc Download(DownloadRequest) returns (DownloadResponse);
+  // DownloadCAS downloads a module in the manifest+blobs encoding format.
+  rpc DownloadCAS(DownloadCASRequest) returns (DownloadCASResponse);
```
Comment (twilly) [inline proto/buf/alpha/registry/v1alpha1/download.proto:27]: Instead of CAS, I'd say blobs or manifest. CAS has no meaning to the caller.

**Comment by bufdev** [comment]
Not blocking this, but added my two cents in the notion doc

**Comment by unmultimedio** [comment]
Closing in favor of https://github.com/bufbuild/buf/pull/1693

**Pull request closed without merging**